### ISSUE TITLE
`QuillEditor` does not rebuild when scrolling

### ISSUE THREAD

**Issue by putnokiabel** [singerdmx/flutter-quill#1907]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### Flutter Quill version

9.3.21

### Steps to reproduce

1. Create a new app with `flutter create quill_test`
2. Replace `main.dart` with the code sample
3. Run the app
4. Note that `QuillEditor` is scrollable
5. Switch to tab 2 by tapping on "2" in the top right corner
6. Note that `QuillEditor` is no longer scrollable

Additionally, note that if in tab 2 you make a scrolling motion (even though the widget does not scroll yet), and then tap on `QuillEditor`, it is now scrolled. This indicates that the `QuillEditor` widget _does_ properly receive the scroll gesture, but the widget does not rebuild (and show the scrolled state) until you tap on it.
If you then go back to tab 1, the same issue occurs. See video demonstration below.

### Expected results

`QuillEditor` is scrollable regardless of the tab, and regardless of whether you tap on the widget or not.

### Actual results

When `QuillEditor` is within a `TabBarView`, it is no longer scrollable after switching tabs.
See video demonstration below.

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

import 'package:flutter_quill/flutter_quill.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Quill scroll bug',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: DefaultTabController(
        length: 2,
        child: Column(
          children: [
            TabBar(
              tabs: [
                Tab(text: '1'),
                Tab(text: '2'),
              ],
            ),
            Expanded(
              child: TabBarView(
                children: [
                  QuillTest(),
                  QuillTest(),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class QuillTest extends StatefulWidget {
  const QuillTest({super.key});

  @override
  State<QuillTest> createState() => _QuillTestState();
}

class _QuillTestState extends State<QuillTest> {
  final QuillController controller = QuillController(
    document: Document()
      ..insert(
        0,
        [for (var i = 0; i < 100; i++) i.toString()].join('\n'),
      ),
    selection: const TextSelection.collapsed(offset: 0),
    readOnly: true,
  );

  @override
  Widget build(BuildContext context) {
    return QuillEditor.basic(
      configurations: QuillEditorConfigurations(
        controller: controller,
        scrollable: true,
        autoFocus: true,
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>
https://github.com/singerdmx/flutter-quill/assets/29568469/d1328ee4-f8bf-469b-9543-77a0226f313f
</details>

### Logs

<details><summary>Logs</summary>

No logs.

</details>

**Comment by salba360496** [comment]
QuillEditor.basic has focusNode property try to use it to set back the focus on quill editor when quill editor is in view.

**Comment by putnokiabel** [comment]
> QuillEditor has focusNode property try to use it to set back the focus on quill editor when it is in view or use autofocus property.

@salba360496 
I've already set `autofocus` to true. I've also tried setting `focusNode` and requesting focus when switching tabs, unfortunately it did not have an effect.

I suspect the problem is not with focus, but something else. Since the widget clearly _does_ receive the scroll gesture, it just doesn't always re-render the widget upon scrolling (until you tap on it). Perhaps a missing `setState` or `notifyListeners` somewhere?

**Comment by CatHood0** [comment]
I resolve this issue using this:

```dart
bool canScroll = false; //set to false to apply just at the first time the setState
Scrollbar(
  notificationPredicate: (notification) {
      if (!canScroll) {
           canScroll = true;
          setState(() {});
      }
      return notification.depth == 0;
  },
  child: <you editor>
),
```

**Comment by putnokiabel** [comment]
> I resolve this issue using this:
> 
> ```dart
> bool canScroll = false; //set to false to apply just at the first time the setState
> //Scrollbar is optional, any widget that uses notificationPredicate for scroll can resolve this (i guess)
> Scrollbar(
>   notificationPredicate: (notification) {
>       if (!canScroll) {
>            canScroll = true;
>           setState(() {});
>       }
>       return notification.depth == 0;
>   },
>   child: <you editor>
> ),
> ```

Thank you @CatHood0 , this workaround seems to fix it! I'll use this for now :)
Would be great to eventually incorporate it into `flutter_quill` so the workaround isn't needed and other people don't run into the issue either!

**Comment by CatHood0** [comment]
@AtlasAutocode do you have idea where can be the issue with the scroll?

This just happens when you use a Scrollbar (in my case).

**Comment by AtlasAutocode** [comment]
I created a project following OP instructions and added the latest Flutter Quill package.
On Windows: Both tabs scroll using mouse wheel and I can switch back and forth with no problems. I can also drag the scrollbar and both tabs scroll. 

I repeated the test on an Android emulator - switching between tabs was difficult because clicks did not register (seem to be obscured by a title bar?) but I could use the tab key to switch - again, the content of both tabs was scrollable using drag action.

One time on windows, however, switching to Tab-2 and using the scroll bar did NOT work. Can't reproduce this so perhaps there is some time-sensitive initialization of something. This would tie into your fix of calling setState to reset/re-initialize the editor state. My guess is the problem is with scrollController but I can't find where or how it is used.

I have found in raw_editor_state.dart:
> late ScrollController _scrollController;

it is set in initState function (L1107)
> _scrollController = widget.configurations.scrollController;

This might not get initialized if something asserted prior to its call and would explain why your fix works.
I would also suggest seeing if you can get the same scroll failure effect when using a release build - I have found that asserts in debug mode don't always get reported (silent failure) but can prevent subsequent code execution and are very difficult to diagnose.

**Comment by putnokiabel** [comment]
> I created a project following OP instructions and added the latest Flutter Quill package. On Windows: Both tabs scroll using mouse wheel and I can switch back and forth with no problems. I can also drag the scrollbar and both tabs scroll.
> 
> I repeated the test on an Android emulator - switching between tabs was difficult because clicks did not register (seem to be obscured by a title bar?) but I could use the tab key to switch - again, the content of both tabs was scrollable using drag action.
> 
> One time on windows, however, switching to Tab-2 and using the scroll bar did NOT work. Can't reproduce this so perhaps there is some time-sensitive initialization of something. This would tie into your fix of calling setState to reset/re-initialize the editor state. My guess is the problem is with scrollController but I can't find where or how it is used.
> 
> I have found in raw_editor_state.dart:
> 
> > late ScrollController _scrollController;
> 
> it is set in initState function (L1107)
> 
> > _scrollController = widget.configurations.scrollController;
> 
> This might not get initialized if something asserted prior to its call and would explain why your fix works. I would also suggest seeing if you can get the same scroll failure effect when using a release build - I have found that asserts in debug mode don't always get reported (silent failure) but can prevent subsequent code execution and are very difficult to diagnose.

@AtlasAutocode
Thank you for looking into it!
Asserts are likely not the issue, since I can reproduce the issue in release builds.

While I can't say anything about Windows (since I don't have a machine that runs Windows), I can reproduce the issue on both iOS and Android with the given code sample.
I've only tested with scrolling by dragging (which is what all mobile users use), not with scrolling by mouse wheel. I can imagine there's a difference there as well.

**Comment by AtlasAutocode** [comment]
Is this still an issue with version 10?
Can you help me reproduce this issue?

**Comment by putnokiabel** [comment]
Yes, I can still reproduce the issue on version `10.3.1`.

Steps to reproduce:
- Create a new app with flutter create quill_test
- Run `flutter pub add flutter_quill` (or add manually to `pubspec.yaml`)
- Replace main.dart with the code sample
- Run the app (I'm running on iOS, but the same issue occurs on Android)
- Note that QuillEditor is scrollable
- Switch to tab 2 by tapping on "2" in the top right corner
- Note that QuillEditor is no longer scrollable

Sample `main.dart`:
```dart
import 'package:flutter/material.dart';

import 'package:flutter_quill/flutter_quill.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Quill scroll bug',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: SafeArea(
        child: DefaultTabController(
          length: 2,
          child: Column(
            children: [
              TabBar(
                tabs: [
                  Tab(text: '1'),
                  Tab(text: '2'),
                ],
              ),
              Expanded(
                child: TabBarView(
                  children: [
                    QuillTest(),
                    QuillTest(),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class QuillTest extends StatefulWidget {
  const QuillTest({super.key});

  @override
  State<QuillTest> createState() => _QuillTestState();
}

class _QuillTestState extends State<QuillTest> {
  final QuillController controller = QuillController(
    document: Document()
      ..insert(
        0,
        [for (var i = 0; i < 100; i++) i.toString()].join('\n'),
      ),
    selection: const TextSelection.collapsed(offset: 0),
    readOnly: true,
  );

  @override
  Widget build(BuildContext context) {
    return QuillEditor.basic(
      controller: controller,
      configurations: const QuillEditorConfigurations(
        scrollable: true,
        autoFocus: true,
      ),
    );
  }
}

```

**Comment by AtlasAutocode** [comment]
Thanks for your detailed guide.
I have been able to reproduce the problem.
I have 'solved' the problem by 2 actions:
1. Double tap in the blank space selects some text, dismiss the popup and the window scrolls.
2. Tap within the text (not the blank space) and the window scrolls.

So, I suspect it has something to do with the editor not being activated correctly. The selection actions must be activating the editor. I will need to dig deeper to try to find out what.

I have also found that if I turn off autoFocus, now any tap within the window will enable scrolling. Perhaps autoFocus is preventing the editor tab being activated? 

Or perhaps there is some problem with flutter's TabController? I am trying to figure out the difference between DefaultTabController and TabController that has a notifier that would enable us to listen to tab changes. That's a deep dive into the inner workings of flutter!

**Comment by putnokiabel** [comment]
Thank you for looking into it @AtlasAutocode ! 

Indeed (right now) you need to tap on the editor to focus it first before it is scrollable.

While listening to tab changes might fix this specific issue (within a tab controller), I believe the fact that the editor needs to be focused is a problem in and of itself, IMO the editor shouldn't need to be tapped in order to scroll.
Just like a `SingleChildScrollView` with a very long `Text` as a child does not need to be tapped to scroll.
It's also nice to be able to scroll an (editable) text field without the keyboard having to pop up.

I believe trying to solve it that way (instead of listening to tab changes) would also solve #2104 and #1841.

**Comment by putnokiabel** [comment]
Also note that the workaround by @CatHood0 seems to work perfectly, so likely it's a case of a missing listener (to scroll notifications) in the scrollable widget of the editor somewhere!

**Comment by AtlasAutocode** [comment]
Just a quick update to say that I am still working on this. These issues are much more complex than I expected and are requiring a deep dive into how flutter implements scrollable widgets.

The quick fix mentioned above works very well for mobile but totally fails for desktop (assert failures).

There are 2 observations so far:
1. scrollController initial offset is in pixels and not lines which is what users would expect.
2. RenderEditor uses QuillSingleChildScrollView which directly creates a Scrollable widget. 
> The flutter documentation states that Scrollable widgets are usually not created de-novo but are a part of ListView or GridView.

When I create a flutter test app using ListView, scrollController works correctly to scroll to the specified position (in pixels!).
Using RenderEditor, scrollController totally ignores any initial position that we set.

Creating a test app with 2 tabs and 2 ListViews or SingleChildScrollViews, scrolling works correctly as expected. Both tabs move to the starting offset and both tabs are scrollable.

Conclusion: the quill editor implementation of scrolling is causing the problem and QuillSingleChildScrollView's handling of Scrollable is not consistent with how ListView handles Scrollable.

I have bypassed QuillSingleChildScrollView and it solves your issue but not the other issues. So, I do not have the correct solution yet, and I do not know whether my solution has any unexpected consequences.
